### PR TITLE
Use HTTP range requests (responses) to serve mp4 from assets

### PR DESCRIPTION
**Pull request by mroswald**
This PR solves a problem when video assets are used from third-party React Native components (e.g. [react-native-video](https://github.com/brentvatne/react-native-video). The video will not work while the assets are served from the react native packager because the used video component (iOS) relies on HTTP range requests.

I added a small fix that allows ranged requests (e.g. mp4) to be served in ranges. 

To test this:

1. make new react native project
1. add [react-native-video](https://github.com/brentvatne/react-native-video) to xcode project
1. add video component to your project
```
import Video from 'react-native-video';
var resolveAssetSource = require('react-native/Libraries/Image/resolveAssetSource');
/* ... /*
render() {
    let source = resolveAssetSource(require('./someVideoFile.mp4')) || {};
    return <Video /*....*/ source={source} />;
}
```

That should not work (if video is smaller than a few megabytes, open app a few times). Then add my fix, that should do the trick.

### CODE REVIEW

**Comment by facebook-github-bot-9** [comment]
By analyzing the blame information on this pull request, we identified @frantic and @martinbigio to be potential reviewers.

**Comment by mroswald** [comment]
relates to brentvatne/react-native-video#246

**Comment by mroswald** [comment]
Not sure if I need to add `mime-types` to `package.json`? It's not a direct dependency of `react-native` yet, but it's in the tree

**Comment by mroswald** [comment]
updated to current master

**Comment by mroswald** [comment]
rebased to current master

File: packager/react-packager/src/Server/index.js [hunk 1]
```diff
@@ -168,6 +169,8 @@ const dependencyOpts = declareOpts({
   },
 });
 
+const isRangeRequest = (req) => req.headers && req.headers.range;
```
Comment (frantic) [inline packager/react-packager/src/Server/index.js]: why not inline it?

File: packager/react-packager/src/Server/index.js [hunk 2]
```diff
@@ -402,7 +405,28 @@ class Server {
     const assetEvent = Activity.startEvent(`processing asset request ${assetPath[1]}`);
     this._assetServer.get(assetPath[1], urlObj.query.platform)
       .then(
-        data => res.end(data),
+        data => {
```
Comment (frantic) [inline packager/react-packager/src/Server/index.js]: I'd extract this function to something like `sendWithRange` module for easier understanding.

**Comment by frantic** [comment]
Great stuff, thank you!

I have a question and a request:

1. Is there already an npm package that solves this? Maybe it's easier to add "serves-content-with-ranges" middleware on higher level? This way anything that packager returns could be requested by range.
2. Mind writing a small unit test for your feature to make sure it doesn't break in the future?

**Comment by mroswald** [comment]
@frantic thanks for your valuable feedback! I checked some npm packages but did nothing that is small enough to fit into React Native's `Server` module. By the way - you know the reason why they did not simply use `express`?

anyway: I added a test and refactored the range serving to a small function which I named `middleware` :-) This is just a proposal, maybe you have a better idea

**Comment by frantic** [comment]
Looks good, thank you!

**Comment by frantic** [comment]
@facebook-github-bot import

**Comment by frantic** [comment]
@facebook-github-bot import

**Comment by facebook-github-bot-7** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1761527644126604/int_phab) to review.

**Pull request closed without merging**